### ISSUE TITLE
Encryption middleware issue when invoking function directly 

### ISSUE THREAD

**Issue by dvuletas** [inngest/inngest-py#185]
Inngest SDK version: 0.4.16 

Hello guys, 
I'm using a new (experimental) feature of Encryption middleware. This works as expected when we use `inngest_client.send(...)`. 
Unfortunately, it does not work when using `step.invoke_by_id()`, it will get stuck in the process of invoking a function by id (probably because that function expects to receive encrypted data and is receiving plain text data)
<img width="1291" alt="Screenshot 2024-12-04 at 13 33 48" src="https://github.com/user-attachments/assets/6b9ac441-966a-4d4c-92be-8c577f108bd5">

When not using encryption middleware, everything is working as expected, and that function passes. 

<img width="483" alt="Screenshot 2024-12-04 at 13 36 31" src="https://github.com/user-attachments/assets/eb961465-e781-4ed4-abb9-548f60d6fc89">

**Comment by JonParton** [comment]
@amh4r - Hi Aron! 

So sorry to ping you on this one, but may be in your Python Wheelhouse 👀 Is this a known issue?

**Comment by amh4r** [comment]
@dvuletas @JonParton, could you describe the setup a little more? Is the parent function using the encryption middleware? Is the child function using it? Are you invoking across languages (e.g. Python -> TypeScript)?

**Comment by dvuletas** [comment]
Hey @amh4r 
So, from the communication side, everything is Python, parent and child. 

Regarding setup:
- We have one instance of a client: 
    ```
     self.client = inngest.Inngest(
                    app_id=service_name,
                    logger=logging.getLogger(__name__),
                    is_production=not self.is_local_or_workflow_env(),
                    middleware=[
                        EncryptionMiddleware.factory(encryption_key, fallback_decryption_keys=[fallback_encryption_key],
                                                     event_encryption_field='sensitive')],
                    signing_key=signing_key,
                    event_key=event_key,
                )
    ``` 
We have one parent function; inside that (parent) function, we invoke multiple other (child) functions.

- Function 1 root function 
   - Function 2 child function
   - Function 3 child function
   - Function 4 child function

When functions 2,3,4 are invoked, we are using `step.invoke_by_id(....) [Step -> inngest.Step]` and function 1 is invoked via `client.send(...)`.

If you need anything more, I can provide more details if needed.

**Comment by amh4r** [comment]
Looking at run `01JE8P3W8SXP2CJT0Y2TRNGDEK`, I noticed that your `Create entry analysis` invoke was successful but something is hanging after your `Submit analysis to 3dat`. Looking into the raw data, I noticed that the latter's payload has a string for its `data` (it looks like stringified JSON).

I get a hanging function when I set `data` to a string:

<details>
<summary>Repro code</summary>

```py
import flask
import inngest
import inngest.flask
from inngest.experimental.encryption_middleware import EncryptionMiddleware


app = flask.Flask("my-app")

inngest_client = inngest.Inngest(
    app_id="my-app",
    middleware=[EncryptionMiddleware.factory("my-secret-key")],
)


@inngest_client.create_function(
    fn_id="child",
    trigger=inngest.TriggerEvent(event="never"),
)
def child(
    ctx: inngest.Context,
    step: inngest.StepSync,
) -> str:
    return f"Hello, {ctx.event.data['name']}!"


@inngest_client.create_function(
    fn_id="parent",
    trigger=inngest.TriggerEvent(event="my-event"),
)
def parent(
    ctx: inngest.Context,
    step: inngest.StepSync,
) -> None:
    step.invoke_by_id(
        "invoke",
        app_id="my-app",
        data="hi",
        function_id="child",
    )


inngest.flask.serve(
    app,
    inngest_client,
    [child, parent],
)

app.run(port=8000)
```
<details>

We need some extra validation since `data` must always be a dictionary. We'll put up a PR with more validation, but in the meantime can you ensure that you aren't setting `data` to a string?

**Comment by dvuletas** [comment]
Hmmm, strange, because when I removed usage of EncryptionMiddleware my client looked like this: 
```python
self.client = inngest.Inngest(
                    app_id=service_name,
                    logger=logging.getLogger(__name__),
                    is_production=not self.is_local_or_workflow_env(),
                    signing_key=signing_key,
                    event_key=event_key,
                )
```
I managed to get a successful response, and everything was executed. 
Run ID: 01JE8SYZQP7AB72HZEWQ40Y52R (without encryption, with the usage of a client in the snippet) Run time: 12/4/2024, 1:22:50 PM

Run ID: 01JE8RTZNEVY53YNZHB42JEMAS (with encryption, with the usage of a client in the snippet from previous comment) Run time: 12/4/2024, 1:03:10 PM

**Comment by amh4r** [comment]
- `01JE8SYZQP7AB72HZEWQ40Y52R` has a dict for the `data` passed to step `Submit analysis to 3dat`.
- `01JE8RTZNEVY53YNZHB42JEMAS` has a string for the `data` passed to step `Submit analysis to 3dat`.

I'm pretty sure that the issue here is that somehow you're passing a string to `data`. Could you open a support ticket where you include the full code for both functions (the parent function and the child function `three_dat_submission`)?

**Comment by dvuletas** [comment]
I played around with this and found one interesting thing. 
- When encryption is enabled, the function `Create entry analysis` returns string type as a response 
  - INFO:app.consumer.workflow.workflow_consumer:Type of entry analysis <class 'str'>
- When encryption is disabled the function `Create entry analysis` returns dict type as a response
  - INFO:app.consumer.workflow.workflow_consumer:Type of entry analysis <class 'dict'> 

Would you please let me know if this helps you? 
If not, I can create a support ticket and add details.

**Comment by amh4r** [comment]
Could you write a minimal reproduction of the bug? I can't replicate it. The following code prints `<class 'dict'>`, so `step.invoke` is properly returning a dict for me.


<details>
<summary>Code</summary>

```py
import flask

import inngest
import inngest.flask
from inngest.experimental.encryption_middleware import EncryptionMiddleware

app = flask.Flask("my-app")

inngest_client = inngest.Inngest(
    app_id="my-app",
    middleware=[EncryptionMiddleware.factory("my-secret")],
)

@inngest_client.create_function(
    fn_id="child",
    trigger=inngest.TriggerEvent(event="never"),
)
def child(
    ctx: inngest.Context,
    step: inngest.StepSync,
) -> dict:
    return {"msg": f"Hello, {ctx.event.data['name']}!"}

@inngest_client.create_function(
    fn_id="parent",
    trigger=inngest.TriggerEvent(event="my-event"),
)
def parent(
    ctx: inngest.Context,
    step: inngest.StepSync,
) -> None:
    out = step.invoke_by_id(
        "invoke",
        app_id="my-app",
        data={"name": "Alice"},
        function_id="child",
    )
    print(type(out))

inngest.flask.serve(
    app,
    inngest_client,
    [child, parent],
)

app.run(port=8000)
```
<details>

**Comment by dvuletas** [comment]
Hey @amh4r I managed to refactor the code, so now we always receive `dict` object. And now it works! 
But, now I have another problem. 

I tried to change your code a bit to validate whether the event in the child function is encrypted, but unfortunately, it is not. Is that expected? I saw that the encryption middleware is present in `step._middleware` but not applied.
Here is the code:

<details>
  <summary>Code</summary>

  ```python
 
    import flask
    import inngest
    import inngest.flask
    from inngest.experimental.encryption_middleware import EncryptionMiddleware
    
    app = flask.Flask("my-app")
    
    inngest_client = inngest.Inngest(
        app_id="my-app",
        middleware=[EncryptionMiddleware.factory("my-secret", event_encryption_field="sensitive")],
        is_production=False
    )
    
    @inngest_client.create_function(
        fn_id="child",
        trigger=inngest.TriggerEvent(event="never"),
    )
    def child(
        ctx: inngest.Context,
        step: inngest.StepSync,
    ) -> dict:
        print(step._middleware.middleware)
        return {"msg": f"Hello, {ctx.event.data['sensitive']['name']}!"}
    
    @inngest_client.create_function(
        fn_id="parent",
        trigger=inngest.TriggerEvent(event="my-event"),
    )
    def parent(
        ctx: inngest.Context,
        step: inngest.StepSync,
    ) -> None:
        print(step._middleware.middleware)
        out = step.invoke_by_id(
            "invoke",
            app_id="my-app",
            data={"age": "10", "sensitive": {"name": "Alice"}},
            function_id="child",
        )
        print(type(out))
    
    inngest.flask.serve(
        app,
        inngest_client,
        [child, parent],
    )
    
    if __name__ == "__main__":
        app.run(port=8000)
   ``` 
<details>
<summary>Results</summary>

```json
{
      "name": "inngest/function.invoked",
      "data": {
        "_inngest": {
          "correlation_id": "01JENR7K11WRA5R4QS81T5C8MR.d6aa1f3a789f3f3108666e0ac807ca5ca7dc5fa5",
          "expire": 1765285361847,
          "fn_id": "my-app-child",
          "gid": "63e60063-c633-4045-a29d-16e7275192b3",
          "name": "invoke",
          "source_app_id": "377bd0d4-b213-5a3a-86ff-0fc9762502ad",
          "source_fn_id": "3b628066-ea3c-5444-91e2-ca4dfa399d87",
          "source_fn_v": 0,
          "tc": {
            "ctx": {
              "sid": "e88d2df2442e50b6",
              "traceparent": "00-7df455ccb307049c970fccfe095fb128-9f499e98e50ddab9-01"
            },
            "ts": "2024-12-09T13:02:41.847371549Z"
          }
        },
        "age": "10",
        "sensitive": {
          "name": "Alice"
        }
      },
      "id": "01JENR7K5QG2A7HA22CS1AH0CG",
      "ts": 1733749361847
}
```
</details>
</details>

**Comment by amh4r** [comment]
@dvuletas, we noticed that bug (invoke event data is not encrypted) while investigating this issue. We merged a fix (#187), but we need to figure out how to release this in a non-breaking way

**Comment by dvuletas** [comment]
Ok, not a problem! 😄 
Please inform me when it will be released.

**Comment by amh4r** [comment]
@dvuletas, we released invoke data encryption in version `0.4.17`. It's opt-in because it's a breaking change, allowing for a zero-downtime release. Our release notes include the guide:
https://github.com/inngest/inngest-py/releases/tag/0.4.17

**Comment by dvuletas** [comment]
Hey @amh4r 
Thanks for your help!
I updated the version of the Inngest to the latest one, and it works. 
I will close this ticket.
